File: apps/a-demo-be-rust-axum/tests/unit/steps/attachment_steps.rs
use cucumber::{given, then, when};

use crate::world::AppWorld;

async fn upload_file(
    world: &mut AppWorld,
    expense_id: uuid::Uuid,
    filename: &str,
    content_type: &str,
    bearer: String,
    data: Vec<u8>,
) {
    world
        .svc_upload_attachment(&bearer, expense_id, filename, content_type, data)
        .await;
}

#[when(
    regex = r#"alice uploads file "([^"]+)" with content type "([^"]+)" to POST /api/v1/expenses/\{expenseId\}/attachments"#
)]
async fn alice_uploads_file(world: &mut AppWorld, filename: String, content_type: String) {
    let bearer = world.bearer();
    let expense_id = match world.last_expense_id {
        Some(id) => id,
        None => return,
    };
    let data = b"fake file content for testing".to_vec();
    upload_file(world, expense_id, &filename, &content_type, bearer, data).await;
}

#[given(regex = r#"alice has uploaded file "([^"]+)" with content type "([^"]+)" to the entry"#)]
async fn alice_uploaded_file(world: &mut AppWorld, filename: String, content_type: String) {
    let bearer = world.bearer();
    let expense_id = match world.last_expense_id {
        Some(id) => id,
        None => return,
    };
    let data = b"fake file content".to_vec();
    upload_file(world, expense_id, &filename, &content_type, bearer, data).await;
}

#[when("alice sends GET /api/v1/expenses/{expenseId}/attachments")]
async fn alice_list_attachments(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = match world.last_expense_id {
        Some(id) => id,
        None => return,
    };
    world.svc_list_attachments(&bearer, expense_id).await;
}

#[when("alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{attachmentId}")]
async fn alice_delete_attachment(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = match world.last_expense_id {
        Some(id) => id,
        None => return,
    };
    let att_id = match world.last_attachment_id {
        Some(id) => id,
        None => return,
    };
    world
        .svc_delete_attachment(&bearer, expense_id, att_id)
        .await;
}

#[when("alice uploads an oversized file to POST /api/v1/expenses/{expenseId}/attachments")]
async fn alice_uploads_oversized(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = match world.last_expense_id {
        Some(id) => id,
        None => return,
    };
    let data = vec![0u8; 11 * 1024 * 1024];
    upload_file(world, expense_id, "big.jpg", "image/jpeg", bearer, data).await;
}

#[then(expr = "the response body should contain {int} items in the {string} array")]
async fn response_array_count(world: &mut AppWorld, count: usize, field: String) {
    let actual = world
        .last_body
        .get(&field)
        .and_then(|v| v.as_array())
        .map(|a| a.len())
        .unwrap_or(0);
    assert_eq!(
        actual, count,
        "Expected {count} items in '{field}', got {actual}, body: {}",
        world.last_body
    );
}

#[then(
    regex = r#"the response body should contain an attachment with "([^"]+)" equal to "([^"]+)""#
)]
async fn attachment_with_field(world: &mut AppWorld, field: String, value: String) {
    let attachments = world
        .last_body
        .get("attachments")
        .and_then(|v| v.as_array());
    let found = attachments
        .map(|arr| {
            arr.iter().any(|att| {
                att.get(&field)
                    .and_then(|v| v.as_str())
                    .map(|s| s == value.as_str())
                    .unwrap_or(false)
            })
        })
        .unwrap_or(false);
    assert!(
        found,
        "Expected attachment with '{field}' = '{value}' in: {}",
        world.last_body
    );
}

#[then("the response body should contain an error message about file size")]
async fn error_file_size(world: &mut AppWorld) {
    let msg = world
        .last_body
        .get("message")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    assert!(
        !msg.is_empty(),
        "Expected file size error, got: {}",
        world.last_body
    );
}

// Bob steps for cross-user ownership tests

#[given(
    regex = r#"bob has created an entry with body \{ "amount": "25\.00", "currency": "USD", "category": "transport", "description": "Taxi", "date": "2025-01-15", "type": "expense" \}"#
)]
async fn bob_created_entry(world: &mut AppWorld) {
    world.svc_login("bob", "Str0ng#Pass2").await;
    if world.last_status == 200 {
        world.bob_auth_token = world
            .last_body
            .get("accessToken")
            .and_then(|v| v.as_str())
            .map(String::from);
        let bob_bearer = world.bob_bearer();
        world
            .svc_create_expense(
                &bob_bearer,
                "25.00",
                "USD",
                "transport",
                "Taxi",
                "2025-01-15",
                "expense",
                None,
                None,
            )
            .await;
        if world.last_status == 201 {
            world.bob_expense_id = world
                .last_body
                .get("id")
                .and_then(|v| v.as_str())
                .and_then(|s| uuid::Uuid::parse_str(s).ok());
        }
    }
}

#[when(
    regex = r#"alice uploads file "([^"]+)" with content type "([^"]+)" to POST /api/v1/expenses/\{bobExpenseId\}/attachments"#
)]
async fn alice_uploads_to_bob_expense(
    world: &mut AppWorld,
    filename: String,
    content_type: String,
) {
    let bearer = world.bearer();
    let expense_id = match world.bob_expense_id {
        Some(id) => id,
        None => return,
    };
    let data = b"file content".to_vec();
    upload_file(world, expense_id, &filename, &content_type, bearer, data).await;
}

#[when("alice sends GET /api/v1/expenses/{bobExpenseId}/attachments")]
async fn alice_list_bob_attachments(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = match world.bob_expense_id {
        Some(id) => id,
        None => return,
    };
    world.svc_list_attachments(&bearer, expense_id).await;
}

#[when("alice sends DELETE /api/v1/expenses/{bobExpenseId}/attachments/{attachmentId}")]
async fn alice_delete_bob_attachment(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = match world.bob_expense_id {
        Some(id) => id,
        None => return,
    };
    let att_id = match world.last_attachment_id {
        Some(id) => id,
        None => return,
    };
    world
        .svc_delete_attachment(&bearer, expense_id, att_id)
        .await;
}

#[when("alice sends DELETE /api/v1/expenses/{expenseId}/attachments/{randomAttachmentId}")]
async fn alice_delete_nonexistent_attachment(world: &mut AppWorld) {
    let bearer = world.bearer();
    let expense_id = match world.last_expense_id {
        Some(id) => id,
        None => return,
    };
    let random_id = uuid::Uuid::new_v4();
    world
        .svc_delete_attachment(&bearer, expense_id, random_id)
        .await;
}

File: apps/demo-be-ts-effect/tests/integration/steps/attachment.steps.ts
import { Given, When, Then } from "@cucumber/cucumber";
import assert from "node:assert/strict";
import type { CustomWorld } from "../world.js";

const MAX_ATTACHMENT_SIZE = 10 * 1024 * 1024; // 10MB

// ---- Given ----

Given(
  /^alice has uploaded file "([^"]*)" with content type "([^"]*)" to the entry$/,
  async function (this: CustomWorld, filename: string, contentType: string) {
    const token = this.tokens.get("alice_access") ?? "";
    const expenseId = this.context["expenseId"] as string;
    const content = Buffer.from("fake file content for testing");
    const res = await this.uploadFile(
      `/api/v1/expenses/${expenseId}/attachments`,
      filename,
      contentType,
      content,
      token,
    );
    if (res.status !== 201) {
      throw new Error(`Failed to upload attachment: ${JSON.stringify(res.body)}`);
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    this.context["attachmentId"] = (res.body as any).id as string;
  },
);

// ---- When ----

When(
  /^alice uploads file "([^"]*)" with content type "([^"]*)" to POST \/api\/v1\/expenses\/\{expenseId\}\/attachments$/,
  async function (this: CustomWorld, filename: string, contentType: string) {
    const token = this.tokens.get("alice_access") ?? "";
    const expenseId = this.context["expenseId"] as string;
    const content = Buffer.from("fake file content for testing");
    this.response = await this.uploadFile(
      `/api/v1/expenses/${expenseId}/attachments`,
      filename,
      contentType,
      content,
      token,
    );
  },
);

When(
  /^alice uploads file "([^"]*)" with content type "([^"]*)" to POST \/api\/v1\/expenses\/\{bobExpenseId\}\/attachments$/,
  async function (this: CustomWorld, filename: string, contentType: string) {
    const token = this.tokens.get("alice_access") ?? "";
    const bobExpenseId = this.context["bobExpenseId"] as string;
    const content = Buffer.from("fake file content for testing");
    this.response = await this.uploadFile(
      `/api/v1/expenses/${bobExpenseId}/attachments`,
      filename,
      contentType,
      content,
      token,
    );
  },
);

When(
  /^alice uploads an oversized file to POST \/api\/v1\/expenses\/\{expenseId\}\/attachments$/,
  async function (this: CustomWorld) {
    const token = this.tokens.get("alice_access") ?? "";
    const expenseId = this.context["expenseId"] as string;
    // Create a buffer just over the limit
    const content = Buffer.alloc(MAX_ATTACHMENT_SIZE + 1, "x");
    this.response = await this.uploadFile(
      `/api/v1/expenses/${expenseId}/attachments`,
      "large.jpg",
      "image/jpeg",
      content,
      token,
    );
  },
);

When(/^alice sends GET \/api\/v1\/expenses\/\{expenseId\}\/attachments$/, async function (this: CustomWorld) {
  const token = this.tokens.get("alice_access") ?? "";
  const expenseId = this.context["expenseId"] as string;
  this.response = await this.get(`/api/v1/expenses/${expenseId}/attachments`, token);
});

When(/^alice sends GET \/api\/v1\/expenses\/\{bobExpenseId\}\/attachments$/, async function (this: CustomWorld) {
  const token = this.tokens.get("alice_access") ?? "";
  const bobExpenseId = this.context["bobExpenseId"] as string;
  this.response = await this.get(`/api/v1/expenses/${bobExpenseId}/attachments`, token);
});

When(
  /^alice sends DELETE \/api\/v1\/expenses\/\{expenseId\}\/attachments\/\{attachmentId\}$/,
  async function (this: CustomWorld) {
    const token = this.tokens.get("alice_access") ?? "";
    const expenseId = this.context["expenseId"] as string;
    const attachmentId = this.context["attachmentId"] as string;
    this.response = await this.delete(`/api/v1/expenses/${expenseId}/attachments/${attachmentId}`, token);
  },
);

When(
  /^alice sends DELETE \/api\/v1\/expenses\/\{bobExpenseId\}\/attachments\/\{attachmentId\}$/,
  async function (this: CustomWorld) {
    const token = this.tokens.get("alice_access") ?? "";
    const bobExpenseId = this.context["bobExpenseId"] as string;
    const attachmentId = this.context["attachmentId"] as string;
    this.response = await this.delete(`/api/v1/expenses/${bobExpenseId}/attachments/${attachmentId}`, token);
  },
);

When(
  /^alice sends DELETE \/api\/v1\/expenses\/\{expenseId\}\/attachments\/\{randomAttachmentId\}$/,
  async function (this: CustomWorld) {
    const token = this.tokens.get("alice_access") ?? "";
    const expenseId = this.context["expenseId"] as string;
    this.response = await this.delete(`/api/v1/expenses/${expenseId}/attachments/nonexistent-id-${Date.now()}`, token);
  },
);

// ---- Then ----

Then(
  "the response body should contain {int} items in the {string} array",
  function (this: CustomWorld, count: number, field: string) {
    assert.ok(this.response !== null);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const body = this.response?.body as Record<string, unknown>;
    const arr = body[field] as unknown[];
    assert.ok(Array.isArray(arr));
    assert.strictEqual(arr.length, count);
  },
);

Then(
  /^the response body should contain an attachment with "([^"]*)" equal to "([^"]*)"$/,
  function (this: CustomWorld, field: string, value: string) {
    assert.ok(this.response !== null);
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const body = this.response?.body as any;
    const attachments = body?.attachments as Array<Record<string, string>>;
    assert.ok(Array.isArray(attachments));
    const found = attachments.some((a) => String(a[field]) === value);
    assert.ok(found);
  },
);

Then("the response body should contain an error message about file size", function (this: CustomWorld) {
  assert.ok(this.response !== null);
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const body = this.response?.body as Record<string, string>;
  const message = (body["message"] ?? body["error"] ?? "").toLowerCase();
  assert.ok(message.length > 0);
});
